SMP Support RISC-V PolarFire SoC Icicle Kit

con-pax opened the pull request:
The first cut of the Risc-V SMP support provides basic support for machine mode SMP operation with integer and optional floating point context saving via the arch_switch mechanism. The configuration options allow selection of a contiguous set of CPUs between 1 and 4 to allow fine grained selection of a subset of the available CPUs. Non-homogeneous CPU setups such as those found in the Microchip PolarFire SOC and Sifive FU540 with a mix of monitor and application CPUs, are supported by allowing the monitor CPU be excluded from the operating set.

petermcs-01 committed:
drivers: timer: adding smp support for riscv_machine_timer

Adding support for smp architectures.
The MTIMECMP is referenced relative to the hart id
to select the correct timer for each hart.

Added smp_timer_init function for secondary hart timer setup

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
arch: riscv: kconfig: Adding IDLE_STACK SIZE

Override the default stack size and test stack size for
64 bit build.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
arch: riscv: include: adding smp macros

Adding rv_smp_defs.h file which  has flags for
wake procedures

Signed-off-by: Peter Mc Shane <peter.mcshane@microchip.com>

petermcs-01 committed:
arch: riscv: include: define CONFIG_USE_SWITCH

arch_thread return value is removed if using switch context
saving.

returning the "nested" value based on the current running
cpu.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
include: arch: arch_inlines support for RISC-V

Adding RISC-V architecture in arch_inlines by including
riscv/arch_inlines.h. This contains macro for arch_curr_cpu.

Add arch_switch wrapper function for RISC-V.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
include: arch: adding structure definition for MTIMECMP register

Added in proper structure for core local interrupt handler as
we needed support for multiple CPUs.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
soc: riscv-privilege: common: add dummy supervisor isr handler

Added dummy supervisor isr handler.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
soc: riscv-privilege: add mpfs soc support

Adding soc suport for the Microchip PolarFire SoC device.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
include: modifying kernel_structs.h

Removed cpu mobile for the moment as it causes recursive lock in risc-v

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
soc: riscv-privilege: common: modifying soc_common_irq.c

changed to arch_irq_lock() as unmatched irq_lock at this stage
leaves global_lock set to 1...

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
arch: riscv: adding config for monitor core

Added configs to allow selection of options for monitor harts,
supervisor mode awareness, starting hart and PMP clearing

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
kernel: add config SMP_BASE_CPU

Added SMP_BASE_CPU to allow selection of contiguous subset
of CPUs for SMP operation.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
riscv: core: fixing caller/callee mismatch in isr.S

Fixup of caller/callee mismatch

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
riscv: core: adding pmp clearing in reset.S

Made PMP clearing configurable

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
riscv: core: simplified macro for HART to use in swap.S

Simplified macro and tagged #else/#endif

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
kernel: config option for range of CPUs in kconfig

Added SMP_BASE_CPU to allow selection of range of CPUs for SMP operation

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
kernel: smp: use correct interrupt stacks for smp riscv

Use correct interrupt stack for SMP setups where we are only
using a subset of the available CPUs.
Also make sure every CPU has a stack even if not being used for Zephyr.
Added CONFIG_MP_TOTAL_CPUS to account for both zephyr and non
Zephyr CPU's in a system.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
riscv: core: isr: fpu support

Moved mepc and mstatus saving up as we may need to modify
mstatus temporarily.
Moved position of macro for selecting switc vs swap as issues
in some of the macro names have been fixed and there is now more
commonality.
Refactored FPU saving to take into account the different relation-
ships of "current" to thread we are interrupting from depending on
interrupt vs sys call etc. Also added/expanded comments to make
this clearer.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
west.yml: changed commit reference for hal_microchip

Pointing to the forked hal_microchip repository in the
polarfire SoC Github organization. A PR is made on the
Official hal_microchip repository

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

edersondisouza commented:
Hi @con-pax,

It's great to see support being added for the PolarFire SoC Icicle Kit! As you know, I've sent #41939 that also touches SMP for RISC-V, but it's qemu only. I hope you've had a chance to take a look at it =D

Let's talk on Zephyr's Discord about ways to converge both efforts? My nick there is also "edersondisouza" - not sure about your timezone, but I'm there usually from 17:00 UTC to 2:00 UTC - but we can certainly find the best time to chat =D

Cheers!

con-pax commented:
Hi @nashif,
Thanks for you feedback, I will reformat the commits, and ensure that the commit messages meet the requirement, then try again.

Hi @edersondisouza, Thanks. that sounds like a great idea! I will come find you on the Discord and we can go from there.

fkokosinski commented on soc/riscv/riscv-privilege/mpfs/mpfs_sys.c:
```diff
@@ -0,0 +1,179 @@
+       MSS_GPIO_enable_irq(GPIO2_LO, MSS_GPIO_31);
+    
+	/* Enable IRQ for PLIC driver */
+	//irq_enable(RISCV_MACHINE_EXT_IRQ);
```
Unused commented-out code?

fkokosinski commented on soc/riscv/riscv-privilege/mpfs/mpfs_sys.h:
```diff
@@ -0,0 +1,106 @@
+    #define SUBBLK_CLOCK_CR_MMUART1_MASK                         (0x01 << 0x6)
+#define SUBBLK_CLOCK_CR_GPIO0_MASK                           (0x01 << 0x14)
+    #define SUBBLK_CLOCK_CR_GPIO2_MASK                           (0x01 << 0x16)
+        #define SUBBLK_CLOCK_CR_CFM_MASK                             (0x01 << 0x1D)
```
Indentation

fkokosinski commented on soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c:
```diff
@@ -0,0 +1,110 @@
+
+#ifdef  MPFS_HAL_HW_CONFIG
+        load_virtual_rom();
+//        config_l2_cache();
```
Unused code

fkokosinski commented on soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c:
```diff
@@ -0,0 +1,110 @@
+//        config_l2_cache();
+#endif  /* MPFS_HAL_HW_CONFIG */
+
+        //init_memory
```
Ditto

fkokosinski commented on soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c:
```diff
@@ -0,0 +1,110 @@
+#ifdef  MPFS_HAL_HW_CONFIG
+       (void)init_bus_error_unit(); //not addded
+//       (void)init_mem_protection_unit();
+//       (void)init_pmp((uint8_t)MPFS_HAl_FIRST_HART);
```
Ditto

fkokosinski commented on soc/riscv/riscv-privilege/mpfs/mpfs_sys1.c:
```diff
@@ -0,0 +1,110 @@
+//        mss_pll_config();
+        
+
+        //PLIC_init_on_reset();
```
Ditto

con-pax commented:
> You need to work on many of those commits and commit messages, squash many of them and reduce this to a small set of commits for each area...

Hi @nashif,

Wondering if I should close this PR, and create a new one from a branch with the rebased/squashed commits?
Kind Regards,
Conor

nashif commented:
> Wondering if I should close this PR, and create a new one from a branch with the rebased/squashed commits?
> Kind Regards,

you do not need to close, you can squash and cleanup commits and force push to the same PR.

petermcs-01 committed:
riscv: core: adding SMP support

Added riscv_smp.c which is used to marshal the startup of
the CPUs. It also includes the ipi support.

Added the offset for gp in the esf structure.
smp support developed on Microchip PolarFire SoC Icicle
Kit

Added arch_switch support for SMP architecture platforms in
thread.c.

Added in arch_switch function to swap.S

reset.s refactored to support marshalling of startup of
multiple CPUs. has support or excluding monitor type cores
with different abilities from those being used in SMP cluster.

isr.s is split into arch_swap and arch_switch implementations,
selected via 'CONFIG_USE_SWITCH'. This is to avoid breaking
existing arch_swap implementations.
This is a basic integer and floating point arch_switch
implementation which does not include support for USERSPACE,
PMP_GUARD and other advanced features.

Dummy handler in _irq_wrapper_s to catch potential spurious
supervisor traps.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
boards: mpfs_icicle: adding PolarFire SoC Icicle Kit support

Adding mpfs_icicle support
Increased DDR memory size as SMP pktqueue sample app would not
fit when stack size was increased.
modules: microchip, adding entry for mpfs
Removed board specific config as it is now covered by arch
level HART_TO_USE.
Added in new macros to configure MPFS buiild correctly
Using macro for total number of cpus available on PolarFire SoC
Icicle Kit.
Select CPU_HAS_FPU_DOUBLE_PRECISION as the libs do not support
single float only for RV64.
Changed architecture to include FD, probably needs overlay or
other mechanism to allow selection at build time but this
is problematic as the DTSI stuff seems to get done at an
earlier stage in the build so I'm not sure how to do it
without overlays.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
include: arch: include general pointer

the gp register is used if compiled with
Medium Low memory model

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
riscv: core: using macro for total cpus in multiprocessor arch

using CONFIG_MP_TOTAL_CPUS to account for both zephyr and
non Zephyr CPUs in system.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
riscv: core: make smp support feature based not device based

Refactored config to make feature based instead of SOC_MPFS
based. Also added macro comments to several #else/#endif
to clarify which condition they belonged to.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
riscv: core: configuring supervisor mode

Allow compile time selection of code to clean up
operation on systems with supervisor mode.

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

petermcs-01 committed:
soc: riscv-privilege: configuring supervisor mode

Allow compile time selection of code to clean up
operation on systems with supervisor mode

Signed-off-by: Conor Paxton <conor.paxton@microchip.com>

kgugala commented on arch/riscv/core/reset.S:
```diff
@@ -58,36 +135,100 @@ boot_master_core:
+	li  t0, MIP_MSIP
+	csrrs x0, mie, t0
 
+loop_slave_core:
```
i don't think we should use `master`/`slave` nomenclature in the new code. Maybe `main`/`secondary` core?

kgugala commented on arch/riscv/core/riscv_smp.c:
```diff
@@ -0,0 +1,219 @@
+void z_riscv_secondary_start(int cpu_num)
+{
+	arch_cpustart_t fn;
+#if 0
+#ifdef CONFIG_SMP
+	z_icache_setup();
+	z_irq_setup();
+
+#endif
+#endif
```
commented out code

kgugala commented on arch/riscv/core/riscv_smp.c:
```diff
@@ -0,0 +1,219 @@
+	riscv_cpu_sp = Z_THREAD_STACK_BUFFER(stack) + sz;
+
+	/* wait slave cpu to start */
+	while (hart_wake_flags[hart_num] != RV_WAKE_WAIT) {
```
what if the secondary core fails to start (for any reason)? Shall we handle such case?

kgugala commented on arch/riscv/core/riscv_smp.c:
```diff
@@ -0,0 +1,219 @@
+#endif
+
+
+#if 0
```
a lot of commented out code, please remove it

kgugala reviewed (changes requested):
looks like this PR is introducing a few changes at once (SMP and new board) making it quite big piece of code to review. Can you split this into smaller PRs - one touching generic RV code and a followup changing the board code?

con-pax commented on arch/riscv/core/reset.S:
```diff
@@ -33,18 +39,89 @@ SECTION_FUNC(reset, __reset)
-	/*
-	 * This will boot master core, just halt other cores.
-	 * Note: need to be updated for complete SMP support
+    la ra, __initialize /* Keeps debugger happy so it does not follow false trails... */
```
> looks like this PR is introducing a few changes at once (SMP and new board) making it quite big piece of code to review. Can you split this into smaller PRs - one touching generic RV code and a followup changing the board code?

Hi @kgugala, thanks for taking the time to look at the pr. Yes, the plan is rebase against latest version and split into smaller pr's.

PiotrZierhoffer reviewed (changes requested):
Some minor remarks - I tried to compile things and failed, and these are the places I had to address locally.

evgeniy-paltsev commented on arch/riscv/core/riscv_smp.c:
```diff
@@ -0,0 +1,219 @@
+		z_arc_connect_ici_clear();
+		IRQ_CONNECT(IRQ_ICI, ARCV2_ICI_IRQ_PRIORITY,
+		    sched_ipi_handler, NULL, 0);
+
+		irq_enable(IRQ_ICI);
+	} else {
+		__ASSERT(0,
+			"ARC connect has no inter-core interrupt\n");
+		return -ENODEV;
+	}
+
+	if (bcr.dbg) {
+	/* configure inter-core debug unit if available */
+		uint32_t core_mask = (1 << CONFIG_MP_NUM_CPUS) - 1;
+		z_arc_connect_debug_select_set(core_mask);
+		/* Debugger halt cores at conditions */
+		z_arc_connect_debug_mask_set(core_mask,	(ARC_CONNECT_CMD_DEBUG_MASK_SH
+			| ARC_CONNECT_CMD_DEBUG_MASK_BH | ARC_CONNECT_CMD_DEBUG_MASK_AH
+			| ARC_CONNECT_CMD_DEBUG_MASK_H));
+
+	}
+
+	if (bcr.gfrc) {
+		/* global free running count init */
+		z_arc_connect_gfrc_enable();
+
+		/* when all cores halt, gfrc halt */
+		z_arc_connect_gfrc_core_set((1 << CONFIG_MP_NUM_CPUS) - 1);
+		z_arc_connect_gfrc_clear();
+	} else {
+		__ASSERT(0,
+			"ARC connect has no global free running counter\n");
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+SYS_INIT(arc_smp_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+#endif
```
This looks like copy-paste of ARC SMP code. Is it really used for risc-V?  :)

con-pax commented on subsys/shell/modules/kernel_service.c:
```diff
@@ -174,9 +174,13 @@ static void shell_stack_dump(const struct k_thread *thread, void *user_data)
 		      size, unused, size - unused, size, pcnt);
 }
 
+if (CONFIG_MP_TOTAL_NUM_CPUS > CONFIG_MP_NUM_CPUS)
```
> 

Hi Piotr, thanks for taking the time to look at the PR. There has been a substantial change to the Zephyr code base, in particular the isr.S over the last couple of days. This PR is becoming quite stale. We have started to break up the work into smaller PR's, starting with board support.

PiotrZierhoffer replied:
Great to hear that! I'd be happy to run it through Renode, so please @ me and @kgugala when you file it!

con-pax commented:
Closing this PR today, as it has gone stale.
PR #44208 has board support for PolarFire SoC Icicle kit and has passed all CI testing.
@kgugala, @PiotrZierhoffer  could I please request you to review it?
Thanks a million for your help so far
